### ISSUE TITLE
Unique index gets droppped and generated each time I use "typeorm migration:generate"

### ISSUE THREAD

**Issue by svegalopez** [typeorm/typeorm#3841]
**Issue type:**

[x ] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[x ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x ] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

1. Add an index to an entity using decorator :
```javascript
@Index(["countryCode", "locode"], { unique:true, where: 'locode IS NOT NULL' })
```
2. use the cli to generate the migration: 
``` 
typeorm migration:generate
```

It comes up with this:
```javascript
await queryRunner.query(`CREATE UNIQUE INDEX "IDX_a4465d440921efae77ef84b626" ON "core"."cities" ("country_code", "locode") WHERE locode IS NOT NULL`);
```

3. Run the migration
```
typeorm migration:run
```

4.  Keep working on your project, add other entities etc.. and then generate a new migration with your latest work:
```
typeorm migration:generate
```

5. You will see this everytime you generate a new migration using the cli:
```javascript
public async up(queryRunner: QueryRunner): Promise<any> {
        await queryRunner.query(`DROP INDEX "IDX_a4465d440921efae77ef84b626" ON "core"."cities"`);
        await queryRunner.query(`ALTER TABLE "core"."products" ADD "name_test" varchar(100) NOT NULL`);
        await queryRunner.query(`CREATE UNIQUE INDEX "IDX_a4465d440921efae77ef84b626" ON "core"."cities" ("country_code", "locode") WHERE locode IS NOT NULL`);
    }
```

As you can see, the UNIQUE INDEX that was added in steps 2 and 3, is getting dropped and generated again and again...

**Comment by smonv** [comment]
Any update on this issue?

**Comment by Hovik123** [comment]
You need to add index on entity and add  { synchronize: false } option for make sure that you disabled the sync for the column.

@Index('customer-id-idx', ['customerId'], { fulltext: true })
@Index('customer-name-idx', ['name'], { fulltext: true })
@Index('customer-address-idx', ['address'], { fulltext: true })
@Index('customer-id-idx', { synchronize: false })
@Index('customer-name-idx', { synchronize: false })
@Index('customer-address-idx', { synchronize: false })
export class CustomerEntity {
}

**Comment by clorichel** [comment]
Thanks Hovik123 for your answer, and svegalopez for detailing this issue in the first place 👍 

---

Facing the same issue with `@Unique` decorators. On an entity:

```typescript
@Unique(['phoneNumber'])
@Unique(['email'])
@Unique(['phoneNumber', 'email'])
```

That created the migration:

```sql
CREATE TABLE myFormerTable, -- and its columns and all
  UNIQUE INDEX `IDX_dc1fffbf22384432b858f905c5` (`phoneNumber`, `email`),
  UNIQUE INDEX `IDX_97672ac88f789774dd47f7c8be` (`email`),
  UNIQUE INDEX `IDX_1e3d0240b49c40521aaeb95329` (`phoneNumber`)
```

And everything is just fine 👌 

Now, adding another entity, which has neither `@Index` nor a `@Unique` nor any relationship with `myFormerTable`, but just a `@PrimaryGeneratedColumn` and some other columns, `typeorm migration:generate` will create the new table just fine 👌 , but also add some weird drop/index actions targeting `myFormerTable` (unchanged) entity:

```sql
DROP INDEX `IDX_dc1fffbf22384432b858f905c5` ON `myFormerTable`;
ALTER TABLE `myFormerTable` ADD UNIQUE INDEX `IDX_97672ac88f789774dd47f7c8be` (`email`);
CREATE UNIQUE INDEX `IDX_dc1fffbf22384432b858f905c5` ON `myFormerTable` (`phoneNumber`, `email`)
```

Weirder, the `IDX_1e3d0240b49c40521aaeb95329` initial index is unchanged when the 2 others are 🙃 

Questions:

1. should I "convert" my `@Unique(["email"])` to `@Index(["email"], { unique: true })` (and my other 2 unique constraints) with the fulltext/synchronize Hovik123's suggestion to avoid my migrations from keeping dropping and recreating indices? My underlying concern being _"I don't wanna break stuff!"_ 😅 
1. should I just not care because it has no impact at all on my database and it's fine dropping and recreating indices?
1. are we sure @imnotjames that the `question` label is appropriate? Sounds like a bug to me 🤔  yet I'm glad if I'm wrong 🤗 

❤️

**Comment by clorichel** [comment]
For the record, and the sake of completeness, I've updated my unique decorators from:

```typescript
@Unique(['phoneNumber'])
@Unique(['email'])
@Unique(['phoneNumber', 'email'])
```

to

```typescript
@Index('user-phonenumber-idx', ['phoneNumber'], { unique: true })
@Index('user-email-idx', ['email'], { unique: true })
@Index('user-phonemail-idx', ['phoneNumber', 'email'], { unique: true })
@Index('user-phonenumber-idx', { synchronize: false })
@Index('user-email-idx', { synchronize: false })
@Index('user-phonemail-idx', { synchronize: false })
```

before `typeorm migration:generate` which did the weirdest thing:

```typescript
await queryRunner.query("DROP INDEX `IDX_1e3d0240b49c40521aaeb95329` ON `users`");
await queryRunner.query("DROP INDEX `IDX_dc1fffbf22384432b858f905c5` ON `users`");
await queryRunner.query("DROP INDEX `IDX_97672ac88f789774dd47f7c8be` ON `users`");
await queryRunner.query("ALTER TABLE `users` ADD UNIQUE INDEX `IDX_1e3d0240b49c40521aaeb95329` (`phoneNumber`)");
await queryRunner.query("ALTER TABLE `users` ADD UNIQUE INDEX `IDX_97672ac88f789774dd47f7c8be` (`email`)");
await queryRunner.query("CREATE UNIQUE INDEX `user-phonemail-idx` ON `users` (`phoneNumber`, `email`)");
await queryRunner.query("CREATE UNIQUE INDEX `user-email-idx` ON `users` (`email`)");
await queryRunner.query("CREATE UNIQUE INDEX `user-phonenumber-idx` ON `users` (`phoneNumber`)");
```

- Dropping former automated indices 👌 
- Creating new manually named indices 👌 
- But what exactly are those `ALTER TABLE` statements? 🙃

**Comment by AlexMesser** [comment]
This was fixed and released. Please update your TypeORM version to the latest and reopen this issue if the error persists.

**Comment by moltar** [comment]
@AlexMesser which version has this fix? Thx

**Comment by thomazmz** [comment]
@AlexMesser I don't think this issue should be closed. I am using TypeORM version 0.2.41 and the issue persists. Please, mention what version does fix the problem or consider reopening the issue. Thanks.

**Comment by smonv** [comment]
@AlexMesser I also think this issue should be open until the fix has been backport to TypeORM 0.2. Our applications also stuck at 0.2 due all breaking changes in 0.3 and waiting for update from nestjs-typeorm

**Comment by eduardomontero** [comment]
I also don't think this should be closed, as I'm using version 0.3.11 and the issue still happens, I've put names to all my indexes, and somehow, there are still 3 indexes that are lost and being dropped and recreated on every migration generation.

**Comment by Plamen5kov** [comment]
@AlexMesser still happens on version: `"typeorm": "^0.3.10"`

**Comment by dwthien** [comment]
@AlexMesser still happens on version: "typeorm": "^0.3.11"
![image](https://user-images.githubusercontent.com/25741737/217136993-01a9534a-43bf-4707-8837-28105212655f.png)

**Comment by DmitriyKhirniy** [comment]
I'm facing same issue.

**Comment by forrestwilkins** [comment]
I'm also facing this issue.

**Comment by gabbrieu** [comment]
any update on this?
3 years have passed and this problem continues

**Comment by metal450** [comment]
We're getting this too, not sure why this was closed when it hasn't really been properly addressed

**Comment by paulochoi** [comment]
For postgres, if you are using named indexes, make sure that these are less than 63 characters, souce:
https://til.hashrocket.com/posts/8f87c65a0a-postgresqls-max-identifier-length-is-63-bytes#:~:text=In%20PostgreSQL%2C%20identifiers%20%2D%2D%20table,the%20allowed%20length%20of%2063.

Changing the index name to be less than that fixed this issue for me.

**Comment by agus-rungie** [comment]
I am still getting this error. Could we reopen this issue? Or did someone find a solution?

**Comment by ninowalker** [comment]
Still a problem 0.3.20.

**Comment by meyvan** [comment]
What works for me is initially when setting up the server I started with ` synchronize: true` then I turned it back off. Then if I try to create a new migration the old values are not showing up

**Comment by chardskarth** [comment]
Does #10041 fix the issue for you @ninowalker ?

**Comment by maisonhai3** [comment]
0.3.20
Postgres.
NestJS.


![Image](https://github.com/user-attachments/assets/8961d25b-8167-4671-8d48-07b96525a7c0)

**Comment by maisonhai3** [comment]
Please reopen this issue. It seems to be existing.

**Comment by mguida22** [comment]
Reopening as it sounds like this is still an issue. If someone can add a minimal reproduction using the latest version of TypeORM that will help us determine a solution.

**Comment by alumni** [comment]
@maisonhai3 Please create a new issue with proper replication steps.

There might be a regression for this issue, but it needs new information. In the meantime, this should stay closed as it was fixed and the tests are passing.
